qualtran/bloqs/gf_arithmetic/gf2_multiplication.py

```diff
@@ -247,3 +247,107 @@ def _gf2_multiplication_symbolic() -> GF2Multiplication:
 _GF2_MULTIPLICATION_DOC = BloqDocSpec(
     bloq_cls=GF2Multiplication, examples=(_gf16_multiplication, _gf2_multiplication_symbolic)
 )
+
+
+@attrs.frozen
+class MultiplyPolyByConstantMod(Bloq):
+    r"""Multiply a polynomial by $f(x)$ modulu $m(x)$. Both $f(x)$ and $m(x)$ are constants.
+
+    Args:
+        f_x: The polynomial to mulitply with, given either a galois.Poly or as
+            a sequence degrees.
+        m_x: The modulus polynomial, given either a galois.Poly or as
+            a sequence degrees.
+
+    Registers:
+        g: The polynomial coefficients (in GF(2)).
+
+    References:
+        [Space-efficient quantum multiplication of polynomials for binary finite fields with
+            sub-quadratic Toffoli gate count](https://arxiv.org/abs/1910.02849v2) Algorithm 1
+    """
+
+    f_x: Poly = attrs.field(converter=lambda x: x if isinstance(x, Poly) else Poly.Degrees(x))
+    m_x: Poly = attrs.field(converter=lambda x: x if isinstance(x, Poly) else Poly.Degrees(x))
+
+    def __attrs_post_init__(self):
+        assert self.m_x.is_irreducible()
+        assert self.f_x.degrees.max() < self.m_x.degrees.max()
+
+    @cached_property
+    def n(self):
+        return self.m_x.degrees.max()
+
+    @cached_property
+    def lup(self) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
+        """Returns the LUP decomposition of the matrix representing the operation.
+
+        If m_x is irreducible, then the operation y := (y*f_x)%m_x can be represented
+        by a full rank matrix that can be decomposed into PLU where L and U are lower
+        and upper traingular matricies and P is a permutation matrix.
+        """
+        n = self.n
+        matrix = np.zeros((n, n), dtype=int)
+        for i in range(n):
+            p = (self.f_x * Poly.Degrees([i])) % self.m_x
+            for j in p.nonzero_degrees:
+                matrix[j, i] = 1
+        P, L, U = GF(2)(matrix).plu_decompose()
+        return np.asarray(L, dtype=int), np.asarray(U, dtype=int), np.asarray(P, dtype=int)
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('g', QBit(), shape=(self.n,))])
```

Comment:
@tanujkhattar I want to have an $n-1$ degree polynomial in GF(2) instead of a number in GF($2^n$). is there a way to do that? or do we want to use GF($2^n$) for consistancy?